### PR TITLE
i18n: add language selection to source/journalist

### PR DESCRIPTION
**Pull request by dachary**
## Status

Ready for review

## Description of Changes

Add language selection to source/journalist interfaces. It will not show by default because there is less than one supported language at the moment. It needs to look better (web design welcome) but the code logic should remain the same.

## Testing

* visual verification can be tried on http://i18n.securedrop.club for both the source and the journalist interface

## Deployment

1. Upgrading existing production instances.

The new configuration variable SUPPORTED_LOCALES is optional and existing installations will not be affected.

2. New installs.

A comment about the SUPPORTED_LOCALES variable is included. An admin can manually define it on the app server for test purposes although it is not documented. Another pull request should implement a way to set this variable during the sdconfig stage, when localization is officially supported.

### If you made changes to the app code:

- [x] Unit and functional tests pass on the development VM

### If you made changes to the system configuration:

- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### CODE REVIEW

File: securedrop/source_templates/index_locales.html [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{% if g.locales|length > 1 %}
```
Comment (heartsucker) [inline securedrop/source_templates/index_locales.html]: I don't like the idea of the `index_locales`. This should just reuse the other `locales` bit and we can use CSS to make it work in both locations. This feels like unnecessarily duplicating code.

File: securedrop/i18n.py [hunk 2]
```diff
@@ -72,3 +75,47 @@ def get_locale():
...
+        return sorted(locales)
+    unsupported = set(supported) - set(locales)
+    if unsupported:
+        raise Exception("config.py SUPPORTED_LOCALES contains {} which is not "
+                        "among the locales found in the {} "
+                        "directory: {}".format(list(unsupported),
+                                               babel.translation_directories,
+                                               locales))
+    if default_locale and default_locale not in supported:
+        raise Exception("config.py SUPPORTED_LOCALES contains {} "
```
Comment (heartsucker) [inline securedrop/i18n.py]: We should make this a `LocaleNotFound` exception or something similar so we can catch it with higher specificity in the unit test.

File: securedrop/tests/test_i18n.py [hunk 3]
```diff
@@ -133,6 +166,9 @@ def test_i18n(self):
         pybabel init -i {d}/messages.pot -d {d} -l zh_Hans_CN
         sed -i -e '/code hello i18n/,+1s/msgstr ""/msgstr "code chinese"/' \
               {d}/zh_Hans_CN/LC_MESSAGES/messages.po
+        pybabel init -i {d}/messages.pot -d {d} -l nb_NO
```
Comment (heartsucker) [inline securedrop/tests/test_i18n.py]: Pedantic: Could you add a newline here since this is a slight wall of text?

File: securedrop/i18n.py (hunk 2 above)
Comment (dachary) [reply securedrop/i18n.py]: Good idea, done !

File: securedrop/tests/test_i18n.py (hunk 3 above)
Comment (dachary) [reply securedrop/tests/test_i18n.py]: Done. Indeed less confusing.

**Comment by dachary** [comment]
modified & repushed. About the index_locale.html I agree that we could DRY this with CSS. However, there is 100%  certainty this will be replaced by something prettier in the immediate future: that's my excuse for not doing it. That being said, if you think it should be done, I'll do it. I don't feel strongly about it.

**Comment by heartsucker** [comment]
I think it's fine if it's ugly for now, and if you remove `index_locales`, the next diff will be cleaner too.

**Comment by dachary** [comment]
@heartsucker removed local_index.html as suggested. It feels cleaner indeed, thanks for pushing in the right direction !

source-index

**Comment by heartsucker** [comment]
@dachary If this gets merged without the prettification of the locales (which I presume it will), I'm leaving it up to you to open a follow up ticket for making it look nice.

**Comment by dachary** [comment]
> I'm leaving it up to you to open a follow up ticket for making it look nice.

Will do :-)

**Comment by redshiftzero** [comment]
Linking to #1966 for `SUPPORTED_LOCALES`

**Pull request merged**